File: GenomeDecoding/awnser.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func decodeGenome(n int, s string) string {
	if n%4 != 0 {
		return "==="
	}

	target := n / 4

	countA, countC, countG, countT, countQ := 0, 0, 0, 0, 0
	for _, ch := range s {
		switch ch {
		case 'A':
			countA++
		case 'C':
			countC++
		case 'G':
			countG++
		case 'T':
			countT++
		case '?':
			countQ++
		}
	}

	if countA > target || countC > target || countG > target || countT > target {
		return "==="
	}

	needA := target - countA
	needC := target - countC
	needG := target - countG
	needT := target - countT

	if needA+needC+needG+needT != countQ {
		return "==="
	}

	result := make([]byte, n)
	for i := 0; i < n; i++ {
		ch := s[i]
		if ch != '?' {
			result[i] = ch
			continue
		}
		if needA > 0 {
			result[i] = 'A'
			needA--
		} else if needC > 0 {
			result[i] = 'C'
			needC--
		} else if needG > 0 {
			result[i] = 'G'
			needG--
		} else if needT > 0 {
			result[i] = 'T'
			needT--
		} else {
			result[i] = 'A'
		}
	}

	return string(result)
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n int
	if _, err := fmt.Fscan(in, &n); err != nil {
		fmt.Fprintln(out, "===")
		return
	}

	var s string
	if _, err := fmt.Fscan(in, &s); err != nil {
		fmt.Fprintln(out, "===")
		return
	}

	if len(s) != n {
		n = len(s)
	}

	result := decodeGenome(n, s)
	fmt.Fprintln(out, result)
}

File: GenomeDecoding/awnser.py
from __future__ import annotations

import sys


def decode_genome(length: int, genome: str) -> str:
    if length % 4 != 0:
        return "==="

    target = length // 4

    count_a = genome.count("A")
    count_c = genome.count("C")
    count_g = genome.count("G")
    count_t = genome.count("T")
    count_q = genome.count("?")

    if (
        count_a > target
        or count_c > target
        or count_g > target
        or count_t > target
    ):
        return "==="

    need_a = target - count_a
    need_c = target - count_c
    need_g = target - count_g
    need_t = target - count_t

    if need_a + need_c + need_g + need_t != count_q:
        return "==="

    result_chars: list[str] = []
    for ch in genome:
        if ch != "?":
            result_chars.append(ch)
            continue

        if need_a > 0:
            result_chars.append("A")
            need_a -= 1
        elif need_c > 0:
            result_chars.append("C")
            need_c -= 1
        elif need_g > 0:
            result_chars.append("G")
            need_g -= 1
        elif need_t > 0:
            result_chars.append("T")
            need_t -= 1
        else:
            result_chars.append("A")

    return "".join(result_chars)


def main() -> None:
    data = sys.stdin.read().strip().split()
    if len(data) < 2:
        print("===")
        return

    length = int(data[0])
    genome = data[1]

    if len(genome) != length:
        length = len(genome)

    decoded = decode_genome(length, genome)
    print(decoded)


if __name__ == "__main__":
    main()
